Preprocess data

ZoeMZou opened the pull request:
Hi @venexia ,

I’ve completed the conversion of preprocess_data.R and modify_dummy_vax_data.R based on the mental health repository. The reason for the action test failure was that I deleted the template dataset-definition.py. I have a couple of points I’m uncertain about and would appreciate your input:

**1. Data Frame Naming:**
I think the data frame should be named "df"  rather than "df_vax" in modify_dummy_vax_data.R to align with how it is called in preprocess_data.R?

 https://github.com/opensafely/post-covid-respiratory/blob/95555dce52f9f759c3c74c84c4b8937be35326c4/analysis/preprocess/modify_dummy_vax_data.R#L7-L13

https://github.com/opensafely/post-covid-mentalhealth/blob/ddb50bd8fc9e4e1a9855dc9f9a03991487c908aa/analysis/modify_dummy_vax_data.R#L8

**2. Timing Code (tic/toc):**
There is no tic() call before the toc() in the respiratory repo. Is this correct? 
https://github.com/opensafely/post-covid-respiratory/blob/95555dce52f9f759c3c74c84c4b8937be35326c4/analysis/preprocess/preprocess_data.R#L158

**3. Death and Deregistration Dates:**
I’ve removed the code for adding death_date and deregistration_date since they are already included in the dataset according to our ehrQL adjustment.

**4. Maybe rebase after merging the branch of dataset definition**
The .yaml file will need to be revised to include actions for generating preprocess data across different cohorts. Maybe can revise after merging the branch for dataset-definition.py?

Thanks for your review. 

Best wishes,
Zoe

venexia commented on analysis/specify_paths.R:
`analysis/specifiy_paths.R` should be in the `.gitignore`. There should be a very similar looking file called `analysis/specifiy_paths_example.R` here instead. 

Note: you won't need this file until after you have released the results so you could remove it from this PR.

ZoeMZou replied:
Have removed the `analysis/specifiy_paths.R`

venexia commented on analysis/preprocess/preprocess_data.R:
```diff
@@ -0,0 +1,157 @@
+
+message("Column names found")
+
+# Identify column classes ------------------------------------------------------
```
Did the variables "has_follow_up_previous_6months", "was_alive", "has_died" and "vax_jcvi_age_*" get renamed in the end? If so, the following needs to be updated.

ZoeMZou replied:
Done

venexia commented on analysis/preprocess/preprocess_data.R:
```diff
@@ -0,0 +1,157 @@
+         contains("out_"), # Outcomes
+         contains("cov_"), # Covariates
+         contains("qa_"), # Quality assurance
+         contains("step"), # diabetes steps
```
You don't need diabetes step variables in this dataset.

ZoeMZou replied:
Done

venexia commented on analysis/preprocess/preprocess_data.R:
```diff
@@ -0,0 +1,157 @@
+
+# Restrict columns and save analysis dataset -----------------------------------
+
+df1 <- df %>% 
```
As above, if the inclusion/exclusion variables were renamed, this needs to be reflected here.

venexia commented on analysis/preprocess/preprocess_data.R:
```diff
@@ -0,0 +1,157 @@
+
+message(paste0("Dataset has been read successfully with N = ", nrow(df), " rows"))
+
+# Format columns ---------------------------------------------------------------
```
Is this necessary given you read the data in with `col_types` specified? If so, it should use the vectors of variables by type rather than re-doing the grep.

ZoeMZou replied:
True, should just use the vectors defined:
https://github.com/opensafely/post-covid-respiratory/blob/52a33bbaaefa71d58cc26648ea9a0f0ac6ea13f4/analysis/preprocess/preprocess_data.R#L65-L74

venexia replied:
Does it change any variables though as you read the data in with col_types specified? It might only be the dates that need updating here.

ZoeMZou replied:
Agree. Other than variables in the vector of date_cols, there is no need to update for `num_`,  `cat_` or `bin_`. 

I have a question here for `_birth_year`, taking `qa_num_birth_year` for example: 
1. We generated it in the variables_cohort.py: `qa_num_birth_year=patients.date_of_birth.year`
2. `qa_num_birth_year`  will be converted to numeric when identifying column classes: 
`num_cols <- c(grep("_num", all_cols, value = TRUE)`
4. These codes will then convert `qa_num_birth_year` to string. 
`          across(contains('_birth_year'), 
                 ~ format(as.Date(., origin = "1970-01-01"), "%Y")), `


Q: do we really use `qa_num_birth_year` as string? or actually numeric as implied from the name?

venexia replied:
`across(contains('_birth_year'),  ~ format(as.Date(., origin = "1970-01-01"), "%Y"))` makes it a date with just the year displayed. I think we probably want to keep this line.

ZoeMZou replied:
Back:
https://github.com/opensafely/post-covid-respiratory/blob/aca510e883769c341c549d86a033a7497b9c2eb9/analysis/preprocess/preprocess_data.R#L66-L70

venexia commented on analysis/preprocess/preprocess_data.R:
```diff
@@ -0,0 +1,157 @@
+  mutate(cov_num_consulation_rate = replace(cov_num_consulation_rate, 
+                                            cov_num_consulation_rate > 365, 365))
+
+# Define COVID-19 severity -----------------------------------------------------
```
We discussed moving this to the study definition. Did you make a decision about whether to do this?

ZoeMZou replied:
Done

venexia commented on analysis/preprocess/preprocess_data.R:
```diff
@@ -0,0 +1,157 @@
+sink()
+message ("Cohort ",cohort_name, " description written successfully!")
+
+# QC for consultation variable and set max to 365 (i.e., one per day) ----------
```
We discussed moving this to the study definition. Did you make a decision about whether to do this?

ZoeMZou replied:
Done

venexia commented on analysis/preprocess/preprocess_data.R:
```diff
@@ -0,0 +1,157 @@
+  cohort_name <- args[[1]]
+}
+
+# Get column names -------------------------------------------------------------
```
Please add `print()` statements for these headings. This will allow us to see how far the script has got if any action should fail.

ZoeMZou replied:
Done

venexia replied:
Hi Zoe - I can't see the print statements, please can you confirm which commit they are in?

venexia reviewed (changes requested):
Hi @ZoeMZou - some further refinements needed, but otherwise looking good.

ZoeMZou commented:
Hi @venexia,

Thank you so much! I’ve completed the revisions based on your comments and the revision we made in the study-definition branch. Shall I rebase now before further revision, so I can:
1. test whether the current codes function correctly with the generated dataset.
2. add the corresponding actions in the YAML file for this branch.

Thanks
Zoe

venexia commented:
Hi @ZoeMZou - yes, please rebase now so we can check the pipeline runs from start to this point successfully. Sorry - I started reviewing and then got to a comment about rebase but will pick this up once you have rebased and sorted any merge conflicts.

ZoeMZou commented:
> Hi @ZoeMZou - yes, please rebase now so we can check the pipeline runs from start to this point successfully. Sorry - I started reviewing and then got to a comment about rebase but will pick this up once you have rebased and sorted any merge conflicts.

Hi @venexia - I have rebased this PR, and the whole pipeline runs successfully from the start to this point. Other than the revision I addressed according to your comments above, I have also did the following revision across scripts to make them run good, of course, some some bugs were detected and fixed:
1. Rename of censoring variables:

- `death_date` to `cens_date_death` throughout all script 
- As we renamed the `inex_` and `cens_`, I have further revised them in the `preprocess.R`, for example, here is more tidy now after deleting those already included in the `purpose_` groups:
https://github.com/opensafely/post-covid-respiratory/blob/632fda9d3faa0931408cdc86f26849bd4bdec157/analysis/preprocess/preprocess_data.R#L91-L107

2. I have no idea whether we would need the `vax_date_` and `vax_cat_` variables as follows in our final dataset, but they were not extracted from the date dataset previously, so I did some codes revision to extract them in:
https://github.com/opensafely/post-covid-respiratory/blob/1daf7f7e36e0a933519b5380221721fe56e647a4/analysis/preprocess/preprocess_data.R#L105-L106

https://github.com/opensafely/post-covid-respiratory/blob/1daf7f7e36e0a933519b5380221721fe56e647a4/analysis/dataset_definition_cohorts.py#L37-L84

ps: we would definitely need `vax_date_eligible` for modifying vax dummy dates in `preprocess.R`, so I extracted it here to make the codes running. 

3. Also spotted the mistake about the name of index dates in each cohort, they should be consistent as  `index_date`, `end_date_exposure`, `end_date_outcome`, previously they were named as for example: `index_prevax` `end_prevax_exposure`, `end_prevax_outcome`
https://github.com/opensafely/post-covid-respiratory/blob/632fda9d3faa0931408cdc86f26849bd4bdec157/analysis/preprocess/preprocess_data.R#L93-L96

Please let me know whether they are right decisions. Happy to revise. 

Best wishes
Zoe

venexia reviewed (changes requested):
Hi @ZoeMZou. This looks good. I think we need to keep the formatting of `qa_num_birth_year` as it was but otherwise, this is ready. To answer your question re the vax variables - these are needed to define the vax and unvax cohorts in the next step of the pipeline.

venexia commented:
[Optional] Also if you are going with a directory approach (like you have for preprocessing) then perhaps you should have one for the dataset definition for consistency?

venexia reviewed (approved):
Hi @ZoeMZou - all approved! Note: I moved utility.R and active_analyses.R back up a directory as utility.R is called across the pipeline and active_analyses.R is the complement to create_project_actions.R. Hope this is okay!

ZoeMZou commented:
> Hi @ZoeMZou - all approved! Note: I moved utility.R and active_analyses.R back up a directory as utility.R is called across the pipeline and active_analyses.R is the complement to create_project_actions.R. Hope this is okay!

Thank you so much, @venexia . Yes, totally make sense. 
Zoe
